docgen/parser/parser.go

```diff
@@ -0,0 +1,213 @@
+package parser
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/golang/protobuf/proto"
+	google_protobuf "github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway/httprule"
+	"github.com/grpc-ecosystem/grpc-gateway/protoc-gen-swagger/options"
+	method "google.golang.org/genproto/googleapis/api/annotations"
+)
+
+const (
+	messageFlag = 4
+	fieldFlag   = 2
+)
+
+// ParseFile parses a proto file.
+func ParseFile(file *google_protobuf.FileDescriptorProto) (*File, error) {
+	comments := parseComments(file.GetSourceCodeInfo())
+
+	pkgName := file.GetPackage()
+	messages := parseMessages(pkgName, comments, file.GetMessageType())
+	services, err := parseServices(pkgName, messages, file.GetService())
+	if err != nil {
+		return nil, err
+	}
+
+	f := &File{
+		Messages: messages,
+		Services: services,
+	}
+
+	if proto.HasExtension(file.GetOptions(), options.E_Openapiv2Swagger) {
+		ext, err := proto.GetExtension(file.GetOptions(), options.E_Openapiv2Swagger)
+		if err != nil {
+			return nil, err
+		}
+		f.Swagger = ext.(*options.Swagger)
+	}
+
+	return f, nil
+}
+
+func parseMessages(pkgName string, comments Comments, messages []*google_protobuf.DescriptorProto) map[string]*Message {
+	res := map[string]*Message{}
+	for i, m := range messages {
+		p := fmt.Sprintf("%d.%d", messageFlag, i)
+		res[getQualifiedName(pkgName, m.GetName())] = &Message{
+			Name:           m.GetName(),
+			Comment:        comments.Get(p),
+			Fields:         parseFields(p, comments, m.GetField()),
+			NestedMessages: parseMessages(pkgName, comments, m.GetNestedType()),
+		}
+	}
+	return res
+}
+
+func parseFields(path string, comments Comments, fields []*google_protobuf.FieldDescriptorProto) []*Field {
+	res := make([]*Field, len(fields))
+	for i, f := range fields {
+
+		res[i] = &Field{
+			Name:    f.GetName(),
+			Comment: comments.Get(fmt.Sprintf("%s.%d.%d", path, fieldFlag, i)),
+			Type:    getType(f),
+		}
+	}
+	return res
+}
+
+func getType(f *google_protobuf.FieldDescriptorProto) string {
+	if t := f.GetTypeName(); t != "" {
+		s := strings.Split(t, ".")
+		return s[len(s)-1]
+	}
+	return strings.ToLower(strings.Replace(f.GetType().String(), "TYPE_", "", 1))
```

Comment:
You probably want to use `TrimPrefix` instead, which is simpler and less prone to confusing behavior